lib/iris/fileformats/_ff.py

```diff
@@ -816,6 +816,35 @@ def __iter__(self):
         return pp._interpret_fields(self._extract_field())
 
 
+def _parse_binary_stream(file_like, dtype=np.float, count=-1):
+    """
+    Replacement `numpy.fromfile` due to python3 performance issues.
+
+    Args:
+
+    * file_like - Standard python file_like object.
+
+    Kwargs:
+
+    * dtype  - Data type to be parsed out, used to work out bytes read in.
+
+    * count - The number of values expected to be generated from the parsing.
+
+    """
+
+    # There are a wide range of types supported, we just need to know the byte
+    # size of the object, so we just make sure we've go an instance of a
+    # np.dtype
+    if not isinstance(dtype, np.dtype):
+        dtype = np.dtype(dtype)
+
+    _buffer = file_like.read(count * dtype.itemsize)
+
+    # Let numpy do the heavy lifting once we've sorted the file reading.
+    array = np.frombuffer(_buffer, dtype=dtype, count=-1)
```

Comment:
@owena11 You're effectively reading the data twice here, right? Once to read the bytes from file, and once to parse the bytes into a `numpy` array using `frombuffer`.

I don't have access to a laptop at the moment, but I'm assuming that a file pointer to a stream doesn't support the buffer protocol? I guess I'm just wondering whether there is a way to do this by only streaming through the data once? 

Perhaps that's not possible...I don't immediately know 🤔

Also, don't we need to care about byte order here?

Reply:
Not sure if I fully understand the first question here, but as an attempt at an answer anyway.....

We _probably_ don't do two passes over the data with this implementation, It's a bit of an guess that I can go and look up the implementation details of numpy if needed, but the `np.frombuffer` has all of the information to create the _header/object_ section of a `np.array`, so the implementation of `frombuffer` probably can be as simple as creating the array object and pointing the internal reference to the data to the `_buffer` object we supply as it supports the buffer protocol, so can share the data.  

This is guess is supported by the associated [flags](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.flags.html#numpy.ndarray.flags) on the array created:
```
  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : False
  WRITEABLE : False
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False
```

The array we create isn't writable because the `bytes` we passed are immutable. Numpy can then handle the copying if we ever need to modify the data (I don't think we ever do modify the header data in this module).

Happily I can answer more confidently for byte order! Byte order will be encoded within the `np.dtype` object, defaulting to the machine default. Throughout the calls in this module the dtype description passed specifies endianness which will be preserved when we convert to the `np.dtype object`.

Reply:
Ah after thinking about this more the arrays generated definitely aren't being modified prior to a copy, otherwise numpy would throw a `ValueError`. However there is no performance hit in my timings for using the mutable `bytearray` rather than simple read so I've push up that as a change.

Would probably prevent annoying gremlins for anyone toughing this code in the future, with more expected behavior.

Reply:
@owena11 Awesome. Okay, let's roll with this... Looks good to me 👍